cache/image_cache.go

```diff
@@ -0,0 +1,90 @@
+package cache
+
+import (
+	"encoding/json"
+	"io"
+	"log"
+
+	"github.com/buildpack/lifecycle"
+	"github.com/buildpack/lifecycle/image"
+	"github.com/pkg/errors"
+)
+
+//go:generate mockgen -package testmock -destination testmock/image_factory.go github.com/buildpack/lifecycle/cache ImageFactory
+
+type ImageFactory interface {
+	NewEmptyLocal(string) image.Image
+}
+
+type ImageCache struct {
+	logger    *log.Logger
+	factory   ImageFactory
+	origImage image.Image
+	newImage  *image.LoggingImage
+}
+
+func NewImageCache(logger *log.Logger, factory ImageFactory, origImage image.Image) *ImageCache {
+	newImage := factory.NewEmptyLocal(origImage.Name())
+	return &ImageCache{
+		logger:    logger,
+		factory:   factory,
+		origImage: origImage,
+		newImage:  image.NewLoggingImage(logger, newImage),
+	}
+}
+
+func (c *ImageCache) Name() string {
+	return c.origImage.Name()
+}
+
+func (c *ImageCache) SetMetadata(metadata lifecycle.CacheMetadata) error {
+	data, err := json.Marshal(metadata)
+	if err != nil {
+		return errors.Wrap(err, "marshall metadata")
+	}
+	return c.newImage.SetLabel(lifecycle.CacheMetadataLabel, string(data))
```

Comment:
If the CacheMetadataLabel name only used in the cache package and is only relevant to one of the caching implementations the label name constant should probably move into the cache package

Reply:
Done.